Standalone setup navigation (#938)
agentgateway/website: assets/agw-docs/standalone/setup/storage.md

> Choose where agentgateway stores the configuration that you manage in the admin UI.

Reviewer: remove that sentence as it is duplicative to the short description

## About

> Agentgateway always reads a configuration file at startup. What the `storage` setting controls is what happens when something *writes* configuration back, such as when you add an MCP server in the admin UI or send a request to the config resource API.

Reviewer: ```suggestion
Agentgateway always reads a configuration file at startup. To control how configuration updates are persisted while agentgateway is running, such as when you use the admin UI or send a request to the config resource API, decide on your storage mode. 
```

### Storage modes

Set the mode in the `config.storage.mode` field of your configuration file. Because the field is in the `config` section, agentgateway applies it at startup only, so a change to it takes effect after a restart.

```yaml
# yaml-language-server: $schema=https://agentgateway.dev/schema/config
config:
  storage:
    mode: hybrid          # file, hybrid, or readOnly
  database:
    url: sqlite:///config/data.db   # required only in hybrid mode
gateways:
  default:
    port: 4000
ui:
  gateways: default
```

Review the following table to understand this configuration.

| Mode | What a write does | Requires a database |
| --- | --- | --- |
| `file` (default) | Agentgateway writes the resource into your configuration file, adding the section if the file does not have one yet. | No |
| `hybrid` | Agentgateway keeps your configuration file as a read-only baseline and stores the resource in the database. At read time, it merges the stored resources over the baseline. | Yes |
| `readOnly` | Agentgateway rejects the write with a `403` response and the message `UI is configured as read-only`. | No |

> In `hybrid` mode, agentgateway never writes back to your configuration file. Instead, it stores the resource in the database and layers it over the file value when it reads the configuration. That is what makes the mode useful for a read-only file, and it is also why the file remains the place to change anything that the UI does not manage, such as a gateway or a listener.

Reviewer: ```suggestion
In `hybrid` mode, agentgateway never writes back to your configuration file. Instead, it stores the resource in the database that you configure in the admin UI or API, and layers it over the file value when it reads the configuration. Note that you can only update certain resources via the admin UI or API. For more information, see [What the database stores](#what-the-database-stores). Updates to gateways or listeners must be made in the configuration file that agentgateway reads at startup. 
```

> | Traffic | Gateways, routes, and TCP routes |

Reviewer: in the previous section, we say that gateways cannot be managed via the UI. but here we say we can. are those different gateways or should the gateway be removed from here?

| UI | Policies and the model catalog |

### How each installation method differs

The mode you want depends on whether your configuration file is writable, which is a property of how you installed agentgateway.

| Method | Configuration file | Default behavior |
| --- | --- | --- |
| [Binary]({{< link-hextra path="/setup/install/binary/" >}}) | A local file, writable. | `file` mode. UI edits are saved to your file. A generated configuration also sets a SQLite database for local runtime features, so `hybrid` mode needs no extra setup. |

> | [Docker]({{< link-hextra path="/setup/install/docker/" >}}) | A mounted file or directory, writable unless you mount it read-only. | `file` mode. UI edits are saved to the file on your host. |

Reviewer: ```suggestion
| [Docker]({{< link-hextra path="/setup/install/docker/" >}}) | A mounted file or directory, writable unless you mount it read-only. | `file` mode. UI edits are saved to the file on your host. `hybrid` mode requires an additional database setup. |
```

> | [Helm]({{< link-hextra path="/setup/install/helm/" >}}) | A ConfigMap that the chart renders from your values and mounts read-only. | The chart sets `file` mode, and because the mount is read-only, a UI save fails. Set the chart's `mode` value to `database` to switch to `hybrid` and store edits in PostgreSQL. |

Reviewer: a little confused if hybrid is an actual setting or if we just called it this way? maybe it would be better to call it database mode?

Reviewer: this table made me also think if all modes are supported for each install method. like for example is there a read-only mode for the binary?

Reviewer: also add a link to the postgres setup further down

   ```

Agentgateway preserves the schema comment at the top of the file, and reloads the file after it writes to it.

> [!IMPORTANT]
> In this mode the UI is a writer of your configuration file, not only a reader. If you keep your configuration in version control, or if you generate it from another tool, use `readOnly` or `hybrid` mode so that a UI edit cannot overwrite it.

### Store UI edits in a database {#binary-hybrid}

Use `hybrid` mode when you want the configuration file to stay exactly as you wrote it, and UI edits to persist somewhere else. Agentgateway accepts a `postgres://` or `postgresql://` URL for PostgreSQL, and treats any other value as a SQLite database path.

1. Set the storage mode and a database URL in your configuration file. A generated configuration already has the `database` field, which points at a SQLite file next to the configuration file. In that case, you add only the `storage` field.

   ```yaml
   # yaml-language-server: $schema=https://agentgateway.dev/schema/config
   config:
     storage:
       mode: hybrid
     database:
       url: sqlite:///config/data.db
   gateways:
     default:
       port: 4000
   ui:
     gateways: default
   ```

> 2. Restart agentgateway. The `config` section is applied at startup, so the new mode does not take effect until the process restarts. For more information, see [Fields that require a restart]({{< link-hextra path="/setup/update/#restart-required" >}}).

Reviewer: there is no command, but maybe that is fine?

3. Confirm the storage mode.

   ```sh
   curl -s http://localhost:15000/api/runtime | jq '.ui.configStoreMode'
   ```

   Example output:

   ```txt
   "hybrid"
   ```

4. Add an MCP server, in the UI or through the config resource API.

   ```sh
   curl -s -X PUT http://localhost:15000/api/config/resources/mcp.target \
     -H 'Content-Type: application/json' \
     -d '{"resources":[{"value":{"name":"persisted-target","mcp":{"host":"http://example.com/mcp"}}}]}'
   ```

   Agentgateway returns the stored resource, with the revision and timestamps that it tracks in the database.

   ```json
   {"resources":[{"kind":"mcp.target","id":"persisted-target","value":{"name":"persisted-target","mcp":{"host":"http://example.com/mcp"}},"revision":1,"createdAt":"2026-08-21T21:51:21.951418797Z","updatedAt":"2026-08-21T21:51:21.951418797Z"}]}
   ```

> 5. Confirm that your configuration file is unchanged, and that the effective configuration includes the server anyway. Agentgateway merges the stored resource over the file.

Reviewer: this step does not really confirm that the file is unchanged. maybe separate these two things?

   ```sh
   curl -s http://localhost:15000/api/config/effective | jq -c '.mcp'
   ```

   Example output:

   ```json
   {"targets":[{"name":"persisted-target","mcp":{"host":"http://example.com/mcp"}}]}
   ```

6. Restart agentgateway again, then confirm that the server is still stored.

   ```sh
   curl -s http://localhost:15000/api/config/resources | jq '.resources[].id'
   ```

   Example output:

   ```txt
   "persisted-target"
   ```

> ### Reject UI edits {#binary-readonly}

Reviewer: maybe we use the modes as H3s? like file, hybrid and read-only?

> Use `readOnly` mode when your configuration file is the only source of truth and you want the UI to be a viewer.

Reviewer: ```suggestion
Use `readOnly` mode when your configuration file is the only source of truth and you want the UI to have read access only.
```

     -w "\nHTTP %{http_code}\n"
   ```

   Example output:

   ```txt
   "UI is configured as read-only"
   HTTP 403
   ```

The UI still shows the running configuration in this mode. Only writes are rejected.

## Helm {#helm}

With the Helm chart, the configuration file is a ConfigMap that the chart renders from your Helm values and mounts read-only at the `/config` path. The proxy reads that file at startup, and the Helm values remain the source of truth.

Read-only storage keeps the deployment reproducible, but it also means that the admin UI cannot save anything. A save returns the following error, because write access to the mounted ConfigMap is denied.

```txt
failed to write to file `/config/config.yaml`: Read-only file system (os error 30)
```

To let the UI store configuration, connect a PostgreSQL instance to your agentgateway pod and switch the Helm chart to database mode. Agentgateway then treats the ConfigMap as a baseline and keeps UI-managed resources in the database. When you make updates to these resources in the UI, agentgateway merges the resources from the database with the baseline in the ConfigMap.

### Chart modes

> | Chart `mode` | Storage mode | Configuration source | UI saves |

Reviewer: is the storage mode the equivalent for the chart mode? I am still a little confused if I need to set both or if the chart mode replaces the storage mode in helm

| --- | --- | --- | --- |
| `readonly` (default) | `file` | The Helm values that you provide to configure agentgateway. The values are translated and stored in a ConfigMap that is mounted to the agentgateway pod. | Config is read-only. UI updates are rejected. |
| `database` | `hybrid` | The ConfigMap as a baseline, with an overlay in PostgreSQL. | Updates to resources that are editable in the UI are stored in the database. |

Everything that you put in your Helm values is part of the baseline, including the fields that you can later edit in the UI. In `database` mode, agentgateway never writes back to the ConfigMap. Instead, it stores your UI edit in the database and layers it over the baseline value when it reads the configuration.

> To choose a mode, set the chart's `mode` value. Do not set the `config.storage.mode` or `config.database.url` fields in your Helm values, because the chart derives both fields from `mode` and overwrites anything that you set for them. The chart also rejects a `mode` value other than `readonly` or `database`.

Reviewer: aah here is the answer :D I would maybe add a sentence to the top that explains that the in Helm installations the storage mode is  defined by the chart mode or mode field. and then in the table, we should maybe say Chart mode and Equivalent storage mode or something along these lines

> [!IMPORTANT]
> Even in database mode, you cannot save the configuration file as a whole in the UI's configuration editor, because the mounted file itself remains read-only. Treat your Helm values as the source of truth for the file, and the UI as the way to manage the resources that are layered on top of it.

### Sections must exist in the ConfigMap

> Adding a capability in the UI adds a section to the configuration file, and in the Helm installation that file is read-only. Although the database can store the resources within a section, the UI cannot add the section itself.

Reviewer: ```suggestion
To allow the UI to configure sections in your configuration file, you must define these sections in your Helm values file, even if they are empty.
```

Without a pre-existing section in the config for MCPs, LLMs, or gateways, the UI navigation shows a **Get started** path, but clicking **Enable** fails with the following error, because enabling the capability requires adding the section to the file.

```txt
File configuration is read-only in hybrid mode. Copy the diff and update the configuration file directly.
```

To manage a capability in the UI, include an empty section for it in your Helm values, as shown in the following example and in the following steps.

```yaml
config:
  mcp:
    targets: []
```

> [!NOTE]
> This constraint is specific to a read-only configuration file. In the binary and Docker installations, the file is writable, so the UI can add a section itself.

### Deploy PostgreSQL

> For a production deployment, use a managed PostgreSQL instance or an operator that handles sbackups and failover. The following example deploys a single instance for testing.

Reviewer: ```suggestion
For a production deployment, use a managed PostgreSQL instance or an operator that handles backups and failover. The following example deploys a single instance for testing.
```
